File: Pinball App 2/Pinball App 2/league/LeaguePreviewParsingSupport.swift
import Foundation

struct LeagueParsedStandingRow {
    let season: Int
    let player: String
    let total: Double
    let rank: Int?
}

struct LeagueParsedStatsRow {
    let season: Int
    let bankNumber: Int
    let player: String
    let machine: String
    let rawScore: Double
    let points: Double
    let eventDate: Date?
    let sourceOrder: Int
}

func resolveLeaguePlayerForStats(preferredPlayer: String?, rows: [LeagueParsedStatsRow]) -> String {
    if let preferredPlayer, !preferredPlayer.isEmpty {
        if rows.contains(where: { leaguePlayerNamesMatch($0.player, preferredPlayer) }) {
            return preferredPlayer
        }
    }

    if let latestRow = rows.max(by: { latestLeagueSortValue($0) < latestLeagueSortValue($1) }) {
        return latestRow.player
    }

    return rows[0].player
}

func latestLeagueSortValue(_ row: LeagueParsedStatsRow) -> Double {
    let dateValue = row.eventDate?.timeIntervalSince1970 ?? 0
    return (dateValue * 1_000_000) + Double(row.season * 100 + row.bankNumber)
}

func latestLeagueSortValue(_ rows: [LeagueParsedStatsRow]) -> Double {
    rows.map(latestLeagueSortValue).max() ?? 0
}

func scopedLeagueStatsRows(_ rows: [LeagueParsedStatsRow], preferredPlayer: String?) -> [LeagueParsedStatsRow] {
    guard let preferredPlayer, !preferredPlayer.isEmpty else { return rows }
    let selectedRows = rows.filter { leaguePlayerNamesMatch($0.player, preferredPlayer) }
    return selectedRows.isEmpty ? rows : selectedRows
}

func leagueAroundRowsWindow(
    rows: [LeagueStandingsPreviewRow],
    selectedIndex: Int,
    windowSize: Int = 5
) -> ArraySlice<LeagueStandingsPreviewRow> {
    guard !rows.isEmpty else { return [] }
    let clampedIndex = max(0, min(selectedIndex, rows.count - 1))
    let edge = windowSize / 2
    let start: Int
    if clampedIndex <= edge {
        start = 0
    } else if clampedIndex >= rows.count - edge - 1 {
        start = max(0, rows.count - windowSize)
    } else {
        start = clampedIndex - edge
    }
    let end = min(rows.count, start + windowSize)
    return rows[start..<end]
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/league/LeaguePreviewParsingSupport.kt
package com.pillyliu.pinprofandroid.league

import com.pillyliu.pinprofandroid.data.leaguePlayerNamesMatch
import java.time.LocalDate
import java.util.Locale

internal fun matchesLeaguePreferredPlayer(candidate: String, preferred: String): Boolean {
    return leaguePlayerNamesMatch(candidate, preferred)
}

internal fun resolveLeaguePlayerForStats(rows: List<StatsCsvRow>, preferredPlayer: String?): String {
    if (!preferredPlayer.isNullOrBlank() && rows.any { matchesLeaguePreferredPlayer(it.player, preferredPlayer) }) {
        return preferredPlayer
    }
    return rows.maxByOrNull(::latestLeagueSortValue)?.player ?: rows.first().player
}

internal fun latestLeagueSortValue(row: StatsCsvRow): Long {
    val datePart = row.eventDate?.toEpochDay() ?: 0L
    return (datePart * 1_000_000L) + (row.season * 100L + row.bank)
}

internal fun scopedLeagueStatsRows(rows: List<StatsCsvRow>, preferredPlayer: String?): List<StatsCsvRow> {
    if (preferredPlayer.isNullOrBlank()) return rows
    val selectedRows = rows.filter { matchesLeaguePreferredPlayer(it.player, preferredPlayer) }
    return if (selectedRows.isEmpty()) rows else selectedRows
}

internal fun normalizeLeagueHeader(value: String): String {
    return value.replace("\uFEFF", "").replace("\u0000", "").trim().lowercase(Locale.US)
}

internal fun coerceLeagueSeason(raw: String): Int {
    val trimmed = raw.trim()
    val digits = trimmed.filter { it.isDigit() }
    return digits.toIntOrNull() ?: trimmed.toIntOrNull() ?: 0
}

internal fun leagueAroundRowsWindow(
    rows: List<StandingsPreviewRow>,
    selectedIndex: Int,
    windowSize: Int = 5,
): List<StandingsPreviewRow> {
    if (rows.isEmpty()) return emptyList()
    val clampedIndex = selectedIndex.coerceIn(0, rows.lastIndex)
    val edge = windowSize / 2
    val start = when {
        clampedIndex <= edge -> 0
        clampedIndex >= rows.size - edge - 1 -> maxOf(0, rows.size - windowSize)
        else -> clampedIndex - edge
    }
    val end = minOf(rows.size, start + windowSize)
    return rows.subList(start, end)
}
